### ISSUE TITLE
[regression] Mate-volume-control doesn’t switch audio streams when changing output device

### ISSUE THREAD

**Issue by geckolinux** [mate-desktop/libmatemixer#3]
Hi,

I'm running the Ubuntu Mate 15.10 Beta release (with Mate 1.10), up to date as of today.  I started a Youtube audio stream through my laptop's default speakers, and then plugged in USB headphones and tried to set them as the default output device using mate-volume-control. However, after selecting them, the Youtube audio kept playing through my laptop speakers. It wasn't until stopping the video and restarting it that the audio came out the selected USB headphones.

I can confirm that this is a regression compared to Ubuntu Mate 15.04, which correctly switches active audio streams as soon as the output device is changed.

Here's the related Launchpad bug report I filed: https://bugs.launchpad.net/ubuntu-mate/+bug/1497107

Thanks for reading!

**Comment by geckolinux** [comment]
Update: I can confirm that this is a regression compared to Ubuntu Mate 15.04, which correctly switches active audio streams as soon as the output device is changed.

**Comment by monsta** [comment]
@mratajsky: can you please take a look at this? for clarification - Ubuntu 15.04 has MATE 1.8 and 15.10 has 1.10.

**Comment by mratajsky** [comment]
The restarting part sounds like there may be ALSA involved, maybe youtube is using a different(ly compiled) flash plugin?

In any case, @sb56637, please confirm it with some regular audio player. If you can confirm it, I'll install 15.10 and see what I can do. I've tried but couldn't reproduce it on my Debian.

**Comment by geckolinux** [comment]
@mratajsky  Thanks for checking into this. I can confirm that this also happens with VLC.

**Comment by zaps166** [comment]
Switching output device sometimes works, sometimes does not (Mate 1.12, Arch/Manjaro Linux).
Automatically after connecting a device Mate doesn't switch to the new device even if PulseAudio plays through it, so volume control affects internal sound device by default. Every connection of Bluetooth/USB device needs that user must open a mixer and switch it manually :/

**Comment by skoehler** [comment]
This is a major annoyance for me. I'm using Mate 1.12 on Gentoo. Changing the output device (e.g. from speaker to USB headset or vice versa) does not affect existing audio streams. The change only takes effect for new audio streams, for example:
1) Open audacious and play a song
2) Change output device (the change doesn't take effect)
3) Hit audacious' stop button
4) Hit the play button (the change now takes effect)

My audacious is definitely using PulseAudio. No ALSA at work here.

**Comment by rhjhunt** [comment]
I can confirm that this is also an issue with Mate 1.12.1 on RHEL 7.2.  Our issue involves pulseaudio, just for information we can change the default sink with the script from:

http://askubuntu.com/questions/71863/how-to-change-pulseaudio-sink-with-pacmd-set-default-sink-during-playback/72076#72076

It would be great to have the functionality returned to do this from the GUI.

**Comment by raveit65** [comment]
New downstream bug report about:
https://bugzilla.redhat.com/show_bug.cgi?id=1315488

**Comment by raveit65** [comment]
Confirmed with fedora 24, but for most applications i need only to restart a playing song.
As a workaround you can use pavucontrol for the moment, here switching the device works out of box during a playing song. Tested with exaile, mate-applel-streamer and kaffeine.

**Comment by jhilliar** [comment]
We intentionally don't load "module-suspend-on-idle" which tends to make this problem worse. I can test a patch on a whole bunch of users pretty easily and quickly if we could get something to try.

**Comment by monsta** [comment]
@mratajsky: so it looks like people experience it with PA, not ALSA. Can you take a look at it?

**Comment by blitzter** [comment]
I am seeing the issue in Mate 1.16.0 + Ubuntu 16.10 when I attach a headphone. Not able to change volume post that using the volume icon or even the media keys(the volume in the volume icon does change)

**Comment by monsta** [comment]
@flexiondotorg: looks like this one is missing Bountysource comment - and people would like to see this fixed... :smile:

**Comment by denis-dordoigne** [comment]
Workaround : after changing parameters, I execute « pulseaudio --kill » : the audio switch to the new output device.

**Comment by xie-dongping** [comment]
Workaround 2:

Set system keyboard short cuts to the following commands:
```
# Volume down
amixer -D pulse sset Master 4%-

# Volume up
amixer -D pulse sset Master 4%+

# Toggle mute
amixer -D pulse set Master 1+ toggle

```

**Comment by dhitchcock** [comment]
Workaround 3:

if you use [indicator-sound-switcher](https://github.com/yktoo/indicator-sound-switcher) the sound moves to the right input, plus it's a lot less steps to switch.  Ignores the bug in the Mate app, but makes life on a laptop plugging and unplugging HDMI audio tolerable.

**Comment by flexiondotorg** [comment]
Something I've had on my list of projects to fund is the inclusion of a "what-did-you-plug-in" plugin, which I think would resolve this issue. Here is the code Ubuntu used http://bazaar.launchpad.net/~unity-settings-daemon-team/unity-settings-daemon/trunk/files/head:/plugins/media-keys/what-did-you-plug-in/

**Comment by geckolinux** [comment]
@flexiondotorg  Hi Martin, that looks nifty. But I wonder if something could be tweaked to make the existing UI work? Because it's not just when I plug in a new device, it's when I switch the default output between different devices that are already connected that this bug happens. It should switch all audio streams to the default output device that I select.

As a quick fix, could `pavucontrol`  be used instead of the Mate sound controls, including when opening audio properties from the panel audio control? That would be a good enough workaround for my usage scenario, because unfortunately I can't really use Mate at the moment due to this bug.

**Comment by raveit65** [comment]
Interesting, for fedora-28 i will go this way.
1. taking ownership of pnmixer (done) https://src.fedoraproject.org/rpms/pnmixer
2. Updating pnmixer (tray-applet) to gtk3 from upstream https://github.com/nicklan/pnmixer
3. using pnmixer-applet as default for f28 and using pavucontrol as default mixer.
preferences of pnmixer:
![pnmixer-pulse](https://user-images.githubusercontent.com/961604/36179371-462ad32c-111c-11e8-8e55-4ab72716c836.png)
A user can replace 'pavucontrol' command with any mixer application you want.
Ie. mate-volume-control, qas-mixer(alsa-mixer), etc.
So pnmixer is high flexible and configurable.

**Comment by raveit65** [comment]
> because unfortunately I can't really use Mate at the moment due to this bug.

omg, please don't use my easy workarround and use another boring desktop 🗡

**Comment by raveit65** [comment]
I have pnmixer with pavucontrol runing here. This will be the new default for fedora 28 Mate-desktop
![pnmixer-with-pavucontrol](https://user-images.githubusercontent.com/961604/36214777-a4a2de66-11aa-11e8-8ab8-c512fc7d1fd6.png)
A few adjustments were needed, feel free to ask for.

**Comment by lukefromdc** [comment]
This no doubt varies with hardware. With the soundcard and headphone jacks on my desktop and with the speakers and headphone jack on my laptop, headphones and speakers seem to be the same device on the digital side, apparently with only a rerouting of the analog audio signal.

EDIT: thus plugging in headphones always just plain works on these boxes

**Comment by geckolinux** [comment]
@lukefromdc  Yes, I should clarify that this bug only exists when switching between different sound cards, such as integrated sound --> USB sound or integrated --> HDMI.

**Comment by monsta** [comment]
Well, looks like we have two issues mentioned here...

The original issue is that after the **manual** switching of output device, the current audio stream (e.g. sound from Youtube, music in audio player) doesn't move to the new device. This worked in MATE 1.8. It used separately built packages for [PulseAudio](https://packages.debian.org/jessie/mate-media-pulse) and [GStreamer](https://packages.debian.org/jessie/mate-media-gstreamer), only one of which could be installed. I assume for most people it was the PA one... In MATE 1.10 a single `mate-media` package replaced them, and this library (`libmatemixer` that is) was created for supporting several backends, including PA and ALSA. We probably have a problem somewhere inside its PA backend.

The other mentioned issue is that the default output device doesn't **auto**-switch from speakers to headphones (Bluetooth or USB) when the latter are plugged in. This is a separate issue, and I have no idea if this ever worked in MATE 1.8 or before.

**Comment by geckolinux** [comment]
@monsta Yep, that sums it up.

Regarding automatically switching active streams to new audio devices that are plugged in, that would indeed be a major refinement. But I don't think that part of the functionality depends on the desktop environment, but rather on a PulseAudio configuration option:
https://wiki.archlinux.org/index.php/PulseAudio#Switch_on_connect

**Comment by monsta** [comment]
I see. So it might just work, independently of libmatemixer.

**Comment by alete89** [comment]
> The original issue is that after the manual switching of output device, the current audio stream (e.g. sound from Youtube, music in audio player) doesn't move to the new device. This worked in MATE 1.8. It used separately built packages for PulseAudio and GStreamer, only one of which could be installed.

this isn't working for me in debian stretch with backported MATE 1.18.0

**Comment by raveit65** [comment]
Does the issue still exists?
